### PR TITLE
Use a HashTable to lookup values or objects (added performance consideration section)

### LINKED ISSUE #9899: Refer to hashtable lookup table
### Prerequisites

- [X] **Existing Issue:** Search the existing issues for this repository. If there is an issue that fits your needs do not file a new one. Subscribe, react, or comment on that issue instead.
- [X] **Descriptive Title:** Write the title for this issue as a short synopsis. If possible, provide context. For example, "Typo in `Get-Foo` cmdlet" instead of "Typo."
- [X] **Verify Version:** If there is a mismatch between documentation and the behavior on your system, ensure that the version you are using is the same as the documentation. Check this box if they match or the issue you are reporting is not version specific.

### Links

https://learn.microsoft.com/en-us/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations

### Summary

The use of a hashtable lookup as defined in [Everything you wanted to know about hashtables](https://learn.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-hashtable) at [As a lookup table](https://learn.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-hashtable#as-a-lookup-table) could be a real game changer if it concerns the performance of a script. Yet it is nowhere mentioned in this document.

### Details

_No response_

### Suggested Fix

It might help others if there would be a "hashtable lookup" section or at least a referral to this the Hashtable document.

### PR DESCRIPTION
**Pull request by iRon7**
# PR Summary
The "Use a HashTable to lookup values or objects" section that I have added to the [PowerShell scripting performance considerations](https://learn.microsoft.com/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations) document shows and explains the impact of using a hashtable to quickly lookup values or objects.

- Fixes #9899

## PR Checklist

- [x] **Descriptive Title:** This PR's title is a synopsis of the changes it proposes.
- [x] **Summary:** This PR's summary describes the scope and intent of the change.
- [x] **Contributor's Guide:** I have read the [contributors guide][contrib].
- [x] **Style:** This PR adheres to the [style guide][style].

[contrib]: https://learn.microsoft.com/powershell/scripting/community/contributing/overview
[style]: https://learn.microsoft.com/powershell/scripting/community/contributing/powershell-style-guide

### CODE REVIEW

**Comment by iRon7** [comment]
I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer

 

Ronald Bode

**Comment by iRon7** [comment]
@microsoft-github-policy-service agree

 

Ronald Bode

File: reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md [hunk 1]
```diff
@@ -204,6 +204,81 @@ finally
 }
 ```
 
+### Looking up entries by property in large collections
+
+It's common to need to use a shared property to identify the same record in different collections,
+like a user name from to retrieve an ID from one list and an email from another. Iterating over the
```
Comment (sdwheeler) [inline reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md:210]:
```suggestion
like using a name to retrieve an ID from one list and an email from another. Iterating over the
```

File: reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md [hunk 2]
```diff
@@ -204,6 +204,81 @@ finally
...
+first list to find the matching record in the second collection is slow. In particular, the
+repeated filtering of the second collection has a large overhead.
+
+Given two collections, one with an **ID** and **Name**, the other with **Name** and **Email:
```
Comment (sdwheeler) [inline reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md:214]:
```suggestion
Given two collections, one with an **ID** and **Name**, the other with **Name** and **Email**:
```

File: reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md [hunk 3]
```diff
@@ -204,6 +204,81 @@ finally
...
+
+```PowerShell
```
Comment (sdwheeler) [inline reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md:216]:
````suggestion
```powershell
````

File: reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md [hunk 4]
```diff
@@ -204,6 +204,81 @@ finally
...
+        Name = "Name$_"
+        Email = "Name$_@fabrikam.com"
+    }
+}
+```
+
+The usual way to reconcile these collections to return a list of objects with the **ID**, **Name**,
+and **Email** properties might look like this:
+
+```PowerShell
```
Comment (sdwheeler) [inline reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md:235]:
````suggestion
```powershell
````

**Review by sdwheeler — CHANGES_REQUESTED** [review]
Just a few nits

**Pull request merged**